### PR TITLE
Add LLaMA scripts

### PR DESCRIPTION
**Pull request by kunal-vaishnavi**
### Description
This PR adds the following scripts for LLaMA:
- LLaMA conversion (support for TorchScript and Dynamo exporters)
- LLaMA parity
- LLaMA benchmark
- LLaMA quantization
- LLaMA integration with [Hugging Face Optimum](https://github.com/huggingface/optimum)

### Motivation and Context
This PR adds scripts for using LLaMA. There is a [follow-up PR](https://github.com/microsoft/onnxruntime/pull/17043) for adding scripts for Whisper.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/models/llama/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+## Benchmark LLaMA
+
+Here are some examples of how you can benchmark LLaMA.
+
+Note: In the below examples, `PyTorch` refers to running in PyTorch without `torch.compile` and `PyTorch 2.0` refers to running in PyTorch with `torch.compile`.
+
+### Variants
+
+1) PyTorch (without `torch.compile`), FP32, Hugging Face `generate()` and `decode()` API
```
Comment (thiagocrepaldi) [inline onnxruntime/python/tools/transformers/models/llama/README.md:60]:
```suggestion
1. PyTorch (without `torch.compile`), FP32, Hugging Face `generate()` and `decode()` API
```
NIT: Markdown style

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+# file with the ".onnx.data" extension.
+def run_dynamo_export(args: argparse.Namespace, l_config: LlamaConfig, llama: LlamaForCausalLM):
+    from torch._dynamo import config
+
+    config.capture_scalar_outputs = True
+
+    # Export decoder_model.onnx
+    input_ids, attn_mask, pos_ids = get_model_inputs(l_config)
+    temp_dir = args.output  # tempfile.TemporaryDirectory()
+    temp_path = os.path.join(temp_dir, "temp.onnx")  # os.path.join(temp_dir.name, "temp.onnx")
```
Comment (thiagocrepaldi) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:145]: Maybe use [NamedTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) would help. When used within a context manager, temp files are automatically deleted for you

You can refer to this [unit test](https://github.com/pytorch/pytorch/blob/5c3aae8385883eee03e42d3a409afcaf1b6e7959/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L846) for an example

**Review by thiagocrepaldi — COMMENTED** [review]
Good job!
If possible, a little more docstrings on main bits and using `NamedTemporaryFile`/`NamedTemporaryDirectory` to prevent accidental leftover files

**Comment by kunal-vaishnavi** [comment]
Splitting changes for LLaMA and Whisper into separate PRs as discussed offline

File: onnxruntime/python/tools/transformers/models/llama/README.md (hunk 1 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/README.md]: Fixed

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 2 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:145]: The temp file would be deleted but I believe the external data files would not be removed. Additionally, until [support for saving large models in ONNX Script is resolved](https://github.com/microsoft/onnxscript/issues/493), the path to save the exported model (`temp_path`) currently has to match the specified path in the ONNX Script code. If a temp file is used, then these paths would not match before running the converter script.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+from benchmark_helper import Precision, prepare_environment, setup_logger  # noqa: E402
+from llama_parity import main as parity_check  # noqa: E402
+from onnx_model import OnnxModel  # noqa: E402
+
+from onnxruntime import quantization  # noqa: E402
+
+logger = logging.getLogger("")
+
+
+def get_model_inputs(config: LlamaConfig):
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:23]: nit: rename as get_sample_inputs

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+    batch_size, seq_len = 2, 8
+    input_ids = torch.randint(low=0, high=config.vocab_size, size=(batch_size, seq_len), dtype=torch.int64)
+    attn_mask = torch.randint(low=0, high=2, size=(batch_size, seq_len), dtype=torch.int64)
+    # pos_ids is of shape (batch_size, seq_len)
+    pos_ids = attn_mask.long().cumsum(-1) - 1
+    pos_ids.masked_fill_(attn_mask == 0, 1)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:29]: pos_ids.masked_fill_(pos_ids< 0, 0)

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+
+from onnxruntime import quantization  # noqa: E402
+
+logger = logging.getLogger("")
+
+
+def get_model_inputs(config: LlamaConfig):
+    batch_size, seq_len = 2, 8
+    input_ids = torch.randint(low=0, high=config.vocab_size, size=(batch_size, seq_len), dtype=torch.int64)
+    attn_mask = torch.randint(low=0, high=2, size=(batch_size, seq_len), dtype=torch.int64)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:26]: nit: Usually we use left side padding for GPT like model. If we want to make it like real input, see example in:
https://github.com/microsoft/onnxruntime/blob/c424e42594d92daba54f264c1c7409e53529d933/onnxruntime/python/tools/transformers/models/gpt2/gpt2_helper.py#L214-L228

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+            raise Exception("Unknown input or output name found")
+    return dynamic_axes
+
+
+def save_onnx_model(onnx_model: onnx.ModelProto, output_path: str, data_path: str):
+    onnx.save(
+        onnx_model,
+        output_path,
+        save_as_external_data=True,
+        all_tensors_to_one_file=True,
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:99]: Is it better to save one tensor per file?

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+    decoder_model_fp32_path = os.path.join(args.output, f"{args.model_name}_decoder_model_fp32.onnx")
+    decoder_with_past_model_fp32_path = os.path.join(
+        args.output, f"{args.model_name}_decoder_with_past_model_fp32.onnx"
+    )
+
+    if args.precision == Precision.FLOAT16:
+        # Convert decoder_model.onnx to FP16
+        decoder_model_fp16_path = os.path.join(args.output, f"{args.model_name}_decoder_model_fp16.onnx")
+        model = OnnxModel(onnx.load_model(decoder_model_fp32_path, load_external_data=True))
+        model.convert_float_to_float16(keep_io_types=False, op_block_list=["If"])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:418]: Generally, we need optimize model first before converting to float16.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+        model.save_model_to_file(
+            decoder_with_past_model_fp16_path, use_external_data_format=True, all_tensors_to_one_file=True
+        )
+        del model
+
+    elif args.precision == Precision.INT8:
+        # Convert decoder_model.onnx to INT8
+        decoder_model_int8_path = os.path.join(args.output, f"{args.model_name}_decoder_model_int8.onnx")
+        quantization.quantize_dynamic(
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:436]: We shall use SmoothQuant for int8 quantization.

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        "-p",
+        "--pytorch",
+        required=False,
+        default=os.path.join("."),
+        help="Directory to PyTorch model and associated files if saved on disk",
+    )
+
+    parser.add_argument(
+        "-o",
+        "--onnxruntime",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/llama_parity.py:107]: onnx_model_directory?

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    parser.add_argument(
+        "-m",
+        "--model_name",
+        required=True,
+        help="Model name in Hugging Face",
+    )
+
+    parser.add_argument(
+        "-p",
+        "--pytorch",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/llama_parity.py:99]: torch_model_directory?

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        action="store_true",
+        help="Print verbose logs",
+    )
+    parser.set_defaults(verbose=False)
+
+    parser.add_argument(
+        "-kv",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/llama_parity.py:131]: `-kv` is not a good abbreviation. May use `-p` here, then use `-t` for torch directory to avoid confliction.

File: onnxruntime/python/tools/transformers/models/whisper/convert_to_onnx.py [hunk 12]
```diff
@@ -183,22 +183,25 @@ def parse_arguments(argv=None):
         "--quantize_embedding_layer",
         required=False,
         action="store_true",
-        help="Produce beam search model with chained encdecinit and decoder.",
+        help="Quantize MatMul, GEMM, and Gather.",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/whisper/convert_to_onnx.py:186]: The change of whisper shall be in another PR.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 3 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py]: Fixed

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 4 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:29]: Those two lines are from Hugging Face's implementation:
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L863-L868

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 5 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py]: This is for exporting so we can use torch.ones here.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 6 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:99]: I set the flag as true by default because it is easier to have one data file vs lots of data files when moving the model and its external data between locations.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 8 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:433]: From this example, SmoothQuant requires [Intel's neural compressor package](https://github.com/intel/neural-compressor) to run. 

https://github.com/microsoft/onnxruntime-inference-examples/blob/main/quantization/language_model/llama/main.py#L260-L272

Is this a dependency that is okay to introduce?

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py (hunk 9 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/llama_parity.py]: This is the path to the ONNX model so `onnx_model_path` might be a better name.

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py (hunk 10 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/llama_parity.py]: Fixed

File: onnxruntime/python/tools/transformers/models/llama/llama_parity.py (hunk 11 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/llama_parity.py]: Fixed

File: onnxruntime/python/tools/transformers/models/whisper/convert_to_onnx.py (hunk 12 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/whisper/convert_to_onnx.py]: Fixed

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 4 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:29]: Those two lines are from Hugging Face's implementation:

https://github.com/huggingface/transformers/blob/7dcd8703ef904adc3ac19b47f769879221c33849/src/transformers/models/llama/modeling_llama.py#L752-L757

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:29]: Those two lines are from [Hugging Face's implementation](https://github.com/huggingface/transformers/blob/7dcd8703ef904adc3ac19b47f769879221c33849/src/transformers/models/llama/modeling_llama.py#L752-L757).

Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:29]: I see. I think you can remove the line since attention_mask is ones, so the line of masked_fill_ has no effect.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 7 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:415]: It's fine that we leave it right now.

We can consider converting model to fp16 first, then run graph optimization so that optimization requires less memory.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 8 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:433]: Yes. Without SmoothQuant, the quality will degrade so it is a must for now.
You can add a requirement.txt file in this directory.

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py]: Added

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 4 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py]: Removed `masked_fill_`

File: onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt [hunk 13]
```diff
@@ -0,0 +1,4 @@
+-r requirements.txt
+--extra-index-url https://download.pytorch.org/whl/cu118
+torch>=2.0.1+cu118
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt:3]: not specify cuda version for package (for example, next version of torch will support cuda 12.*).

File: onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+-r requirements.txt
+--extra-index-url https://download.pytorch.org/whl/cu118
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt:2]: `--extra-index-url` will trigger security warnings later.

File: onnxruntime/python/tools/transformers/models/llama/README.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+# From wheel:
+$ python3 -m onnxruntime.transformers.models.llama.convert_to_onnx -m meta-llama/Llama-2-7b-hf --output llama2-7b
+```
+
+To make this option compatible with [Hugging Face's Optimum](https://github.com/huggingface/optimum), you will need to create `config.json` and `generation_config.json` for your model and store them in the same directory as your ONNX models. For example, you can find those JSON files for LLaMA-2 7B on Hugging Face [here](https://huggingface.co/meta-llama/Llama-2-7b-hf).
+
+Option 2: from [Microsoft's custom export](https://github.com/microsoft/Llama-2-Onnx)
+
+Option 3: from [Hugging Face's Optimum](https://github.com/huggingface/optimum)
+
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/README.md:23]: example command line of optimum.

File: onnxruntime/python/tools/transformers/models/llama/README.md [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+$ python3 convert_to_onnx.py -m meta-llama/Llama-2-7b-hf --output llama2-7b --precision fp16
+
+# From wheel:
+$ python3 -m onnxruntime.transformers.models.llama.convert_to_onnx -m meta-llama/Llama-2-7b-hf --output llama2-7b --precision fp16
+```
+
+INT8:
+```
+# From source:
+$ python3 convert_to_onnx.py -m meta-llama/Llama-2-7b-hf --output llama2-7b --precision int8 --smooth_quant
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/README.md:51]: when the output folder is same for fp32 and int8. This will remove previously exported fp32 models. I suggest to use different folder name for each data type like `--output llama2-7b-int8`

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/README.md:51]: Both the FP32 and INT8 models are created and saved in the output folder when running this command. The script removes the models so that the data file doesn't double in size if the same export command is run multiple times. I'll change the folder names for each data type.

File: onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt (hunk 13 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt]: Changed to comment so the user installs the desired package manually

File: onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt (hunk 14 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/requirements-cuda.txt]: Removed

File: onnxruntime/python/tools/transformers/models/llama/README.md (hunk 15 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/README.md:45]: Added

File: onnxruntime/python/tools/transformers/models/llama/__init__.py [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation.  All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import os
+import sys
+
+sys.path.append(os.path.dirname(__file__))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/__init__.py:8]: Please update the file like 
https://github.com/microsoft/onnxruntime/blob/f044b14e73a7adddc5ebadeaace75f434db8a6a8/onnxruntime/python/tools/transformers/models/whisper/__init__.py

Then you do not need update sys.path in other .py files in this directory.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+    for name in input_names + output_names:
+        if name in input_names:
+            # shape is (batch_size, sequence_length)
+            dynamic_axes[name] = {0: "batch_size", 1: "sequence_length"}
+        elif name == "logits":
+            # shape is (batch_size, sequence_length, vocab_size)
+            dynamic_axes[name] = {0: "batch_size", 1: "sequence_length"}
+        elif "present" in name:
+            # shape is (batch_size, num_heads, past_sequence_length + 1, head_size)
+            dynamic_axes[name] = {0: "batch_size", 2: "past_sequence_length + 1"}
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:35]: For model without past input, I think present shape is  (batch_size, num_heads, **sequence_length**, head_size)

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+                decoder_with_past_model_fp32_path,
+                decoder_with_past_model_int8_path,
+                op_types_to_quantize=["MatMul", "Gemm", "Gather"]
+                if args.quantize_embedding_layer
+                else ["MatMul", "Gemm"],
+                per_channel=args.quantize_per_channel,
+                reduce_range=args.quantize_reduce_range,
+                use_external_data_format=True,
+                extra_options={"MatMulConstBOnly": True},
+            )
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:562]: what if neither --smooth_quant nor --quantize_dynamic is specified.

I suggest to merge them into one argument, like --quantization_method, choice is ["smooth_quant", "quantize_dynamic"], and default value is "smooth_quant". In the future, we will also need support int4 quantization methods.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 20]
```diff
@@ -0,0 +1,579 @@
...
+
+    smooth_quant_group.add_argument(
+        "--nc_workspace",
+        required=False,
+        type=str,
+        default=os.path.join(".", "nc_workspace"),
+        help="Workspace to save intermediate files generated by Intel's Neural Compressor package.",
+    )
+
+    quantize_dynamic_group = parser.add_argument_group("quantize_dynamic")
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:358]: I do not think dynamic quantization is useful for LLAMA. I suggest to remove it.

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:358]: I also see precision loss. Deprecating it in favor of SmoothQuant.

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py (hunk 19 above)
Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:559]: Added `--quantization_method` and deprecated `quantize_dynamic`. For INT4 quantization, we can use the `--precision` flag.

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:559]: Added `--quantization_method` and deprecated `quantize_dynamic`. For INT4 quantization, we can use the `--precision` flag and specify the method with `--quantization_method`.

File: onnxruntime/python/tools/transformers/models/llama/benchmark.py [hunk 21]
```diff
@@ -0,0 +1,524 @@
...
+    setup_logger(args.verbose)
+    logger.info(args.__dict__)
+    torch.backends.cudnn.benchmark = True
+
+    tokenizer = LlamaTokenizer.from_pretrained(args.model_name)
+    config = LlamaConfig.from_pretrained(args.model_name)
+    target_device = f"cuda:{args.device_id}" if args.device != "cpu" else args.device
+    use_fp16 = args.precision == "fp16"
+
+    setattr(args, "tokenizer", tokenizer)  # noqa: B010
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/benchmark.py:507]: instead of passing through args, we can create a data class to pass information.

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py [hunk 22]
```diff
@@ -0,0 +1,576 @@
...
+        "--smooth_quant_dataset",
+        required=False,
+        default="NeelNanda/pile-10k",
+        help="Path to dataset for calibration during quantization",
+    )
+
+    smooth_quant_group.add_argument(
+        "--pad_max",
+        required=False,
+        default=196,
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:334]: why the default value is 196?

Comment (kunal-vaishnavi) [reply onnxruntime/python/tools/transformers/models/llama/convert_to_onnx.py:334]: The SmoothQuant implementation is based on [this example](https://github.com/microsoft/onnxruntime-inference-examples/blob/main/quantization/language_model/llama/main.py) and the default value for `pad_max` is used [here](https://github.com/microsoft/onnxruntime-inference-examples/blob/07c4d037541f74327b8bf10a162db6c4e9bfdbd7/quantization/language_model/llama/main.py#L90-L94).
